rider/backend.gradle

```diff
@@ -5,15 +5,99 @@ import org.apache.tools.ant.taskdefs.condition.Os
 import java.util.regex.Matcher
 import java.util.regex.Pattern
 
+ext.isWindows = Os.isFamily(Os.FAMILY_WINDOWS)
+ext.backend = new BackendHelper("..", logger)
+
+ext.msbuild = new File("")
+ext.artifactDefs = []
+
+class MSBuildTask extends DefaultTask {
+    @InputFile
+    def File buildFile
+
+    // Set msbuild verbosity based on gradle log level
+
+    @TaskAction
+    def build() {
+        project.backend.ReportProgress("Building $buildFile ($project.BuildConfiguration)")
+
+        project.exec {
+            executable = findMSBuildPath()
+            args = [ buildFile, "/p:Configuration=$project.BuildConfiguration;Version=$project.version", "/v:minimal", "/nologo" ]
+        }
+    }
+
+    def findMSBuildPath() {
+        if (project.ext.has('msbuildPath')) {
+            logger.lifecycle("msbuildPath (cached): $project.msbuildPath")
+            return project.msbuildPath
+        }
+
+        project.ext.msbuildPath = project.isWindows ? findMSBuildPathWindows() : findMSBuildPathUnix()
+        logger.lifecycle "msbuildPath: $project.msbuildPath"
+        return project.msbuildPath
+    }
+
+    def findMSBuildPathWindows() {
+        def stdout = new ByteArrayOutputStream()
+        project.exec {
+            executable = project.file(backend.repositoryRoot, "tools\\vswhere.exe")
+            assert executable.isFile()
+            args = [ "-latest", "-products", "*", "-requires", "Microsoft.Component.MSBuild", "-property", "installationPath" ]
+            standardOutput = stdout
+        }
+
+        def buildToolsDir = project.file(stdout.toString().trim())
+        assert buildToolsDir.isDirectory()
+
+        return project.file(buildToolsDir, 'MSBuild\\15.0\\Bin\\MSBuild.exe')
+    }
+
+    def findMSBuildPathUnix() {
+        def stdout = new ByteArrayOutputStream()
+        project.exec {
+            executable = 'which'
+            args = [ 'msbuild' ]
+            standardOutput = stdout
+        }
+
+        return project.file(stdout.toString().trim())
+    }
+}
+
+class RestorePackagesTask extends DefaultTask {
+    @InputFile
+    def File packagesFile
+
+    @TaskAction
+    def restore() {
+        if (project.SkipNuGetRestore) {
+            logger.lifecycle('SkipNuGetRestore=true, skipping restore')
+            return
+        }
+
+        project.backend.ReportProgress("Restoring packages for $packagesFile")
+
+        // TODO: What is Source?
```

Comment:
`Source` is always set on CI and it is the directory which contains the Rider SDKs that the plugin is being built against.